Add EIP: NONREENTRANT and REENTRANT opcodes (#8543)
ethereum/EIPs: EIPS/nonreentrant.md

> ---

Reviewer: ```suggestion
--- 
eip: 7705
```
Assigning next sequential EIP/ERC/RIP number.

Please also update the filename.

title: NONREENTRANT and REENTRANT opcodes
description: Opcodes to mark a contract as nonreentrant
author: Charles Cooper (@charles-cooper)

> discussions-to: https://ethereum-magicians.org/t/eip-for-nonreentrant-opcodes/19957

Reviewer: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-7705-nonreentrant-opcodes/19957
```

status: Draft
type: Standards Track
category: Core
created: 2024-05-09
---

## Abstract

Add two opcodes, `NONREENTRANT` and `REENTRANT`, which set and clear a contract's reentrancy status. After invoking `NONREENTRANT`, a contract cannot be `CALL`ed (or `STATICCALL`ed, or `DELEGATECALL`ed) until `REENTRANT` is invoked.

## Motivation

> Reentrancy attacks account for a substantial portion of user funds stolen on EVM chains, including the famous "DAO hack" (a more complete list is available at https://github.com/pcaversaccio/reentrancy-attacks). However, due to the cost of preventing against reentrancy attacks in application code, developers often opt-out of reentrancy protection. This cost has come down with the advent of transient storage (EIP-1155), but it is still not cheap enough where it is a "no-brainer" to use it by default. This EIP proposes opcodes which make it cheaper to protect against reentrancy in application code.

Reviewer 2: ```suggestion
Reentrancy attacks account for a substantial portion of user funds stolen on EVM chains, including the famous "DAO hack" (a more complete list is available at https://github.com/pcaversaccio/reentrancy-attacks). However, due to the cost of preventing against reentrancy attacks in application code, developers often opt-out of reentrancy protection. This cost has come down with the advent of transient storage ([EIP-1153](./eip-1153.md)), but it is still not cheap enough where it is a "no-brainer" to use it by default. This EIP proposes opcodes which make it cheaper to protect against reentrancy in application code.
```

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> Two new opcodes are introduced, `NONREENTRANT` (0xF6) and `REENTRANT` (0xF7), which set and clear a contract's nonreentrancy flag. The effect of invoking `NONREENTRANT` is that a contract can no longer have execution context transferred to it (via any of the `*CALL` opcodes), until `REENTRANT` is invoked. `CALL`ing a contract which has the nonreentrancy flag set is equivalent to executing a single `REVERT` opcode.

Reviewer 2: FYI, `0xf6` is `AUTH` and `0xF7` is `AUTHCALL` if EIP-3074 makes it...
